File: self/day13/trapped.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

var (
	// key depth value range
	dr       = make(map[int]int)
	maxlevel int
)

// FWScanner is the firewall scanner
type FWScanner struct {
	depth, srange, position, direction int
}

// NewFWScanner blah blah
func NewFWScanner(depth, srange int) *FWScanner {
	return &FWScanner{depth: depth,
		srange:    srange,
		position:  0,
		direction: 1}
}

func (f FWScanner) copy() *FWScanner {
	return &FWScanner{depth: f.depth,
		srange:    f.srange,
		position:  f.position,
		direction: f.direction}
}

func (f *FWScanner) move() {
	if f.position < f.srange {
		f.position += f.direction
	}
	if f.position == 0 {
		f.direction = 1
	} else if f.position == f.srange-1 {
		f.direction = -1
	}
}

func (f FWScanner) severity() int {
	return f.depth * f.srange
}

func (f FWScanner) String() string {
	var retlist []string
	for i := 0; i < f.srange; i++ {
		if i == f.position {
			retlist = append(retlist, "S")
		} else {
			retlist = append(retlist, " ")
		}
	}

	return fmt.Sprintf("[%s]", strings.Join(retlist, " "))
}

// Firewall isn't a sparse array of FWScanners
type Firewall []*FWScanner

// NewFirewall blah blah
func NewFirewall() Firewall {
	var firewall = make([]*FWScanner, maxlevel+1)
	for i := 0; i <= maxlevel; i++ {
		if r, ok := dr[i]; ok {
			firewall[i] = NewFWScanner(i, r)
		}
	}
	return firewall
}

func (fw Firewall) copy() Firewall {
	var dst = make([]*FWScanner, len(fw))
	for idx, fws := range fw {
		if fws != nil {
			dst[idx] = fws.copy()
		}
	}
	return dst
}

func (fw Firewall) String() string {
	var retlist []string

	for _, fws := range fw {
		if fws == nil {
			retlist = append(retlist, "nil")
		} else {
			retlist = append(retlist, fws.String())
		}
	}

	return fmt.Sprintf("[%s]", strings.Join(retlist, ", "))
}

// if go had sum types, i'd return something like
// datatype severityorcollision = Severity of int | Collision of bool
// alas.
func (fw Firewall) advance(packetlevel int, reportcollision bool) int {
	severity := 0
	for idx, fws := range fw {
		if fws != nil {
			if packetlevel == idx && fws.position == 0 {
				if reportcollision {
					return 1
				}
				severity += fws.severity()
			}
			fws.move()
		}

	}
	return severity
}

func (fw Firewall) traverse(packetlevel int, reportcollision bool) (int, int) {
	retval := fw.advance(packetlevel, reportcollision)

	return retval, packetlevel + 1
}

func one() int {
	firewall := NewFirewall()
	var packetlevel, severity, s int

	for packetlevel <= maxlevel {
		s, packetlevel = firewall.traverse(packetlevel, false)
		severity += s
	}
	return severity
}

func two() int {
	var packetlevel, delay, s int
	var failed bool
	firewall := NewFirewall()
	var saved = firewall.copy()

	for {
		packetlevel = 0
		failed = false

		if delay > 0 {
			firewall.advance(-1, false)
			saved = firewall.copy()
		}

		for packetlevel <= maxlevel {
			s, packetlevel = firewall.traverse(packetlevel, true)
			if s == 1 {
				failed = true
				if delay%100000 == 0 {
					log.Printf("failed at delay %d", delay)
				}
				break
			}
		}

		if !failed {
			return delay
		}
		firewall = saved.copy()
		delay++
	}
}

func main() {
	var err error

	flag.Parse()
	if flag.NArg() == 0 {
		log.Fatal("missing input filename")
	}

	infile, err := os.Open(flag.Arg(0))
	if err != nil {
		log.Fatal(err)
	}

	defer infile.Close()

	var d, r int

	scanner := bufio.NewScanner(infile)
	for scanner.Scan() {
		parts := strings.Split(scanner.Text(), ": ")
		if d, err = strconv.Atoi(parts[0]); err != nil {
			log.Fatal(err)
		}
		if r, err = strconv.Atoi(parts[1]); err != nil {
			log.Fatal(err)
		}
		dr[d] = r
		if d > maxlevel {
			maxlevel = d
		}
	}

	log.Println(one())
	log.Println(two())
}

File: 2017/self/day13/trapped.py
#!/usr/bin/env python

from copy import deepcopy
from pprint import pprint
import sys


class Scanner():
    def __init__(self, depth, range_):
        self.depth = depth
        self.range = range_
        self.reset()
    def reset(self):
        self.position = 0
        # 1 down, -1 up
        self.direction = 1
    def move(self):
        if self.position < self.range:
            self.position += self.direction
        if self.position == 0:
            self.direction = 1
        elif self.position == self.range - 1:
            self.direction = -1
    def severity(self):
        if self.position == 0:
            return self.depth * self.range
        return 0
    def __repr__(self):
        ls = [" "] * self.range
        ls[self.position] = "S"
        return str(ls)


def advance(firewall, packetlevel, reportcollision=False):
    severity = 0
    for i in range(len(firewall)):
        if firewall[i]:
            if packetlevel == i and firewall[i].position == 0:
                if reportcollision:
                    return True
                severity += firewall[i].severity()
            firewall[i].move()

    return severity


def newfirewall(dr):
    maxlevel = max(dr.keys())
    firewall = []
    for i in range(maxlevel+1):
        if i in dr:
            firewall.append(Scanner(i, dr[i]))
        else:
            firewall.append([])

    return maxlevel, firewall


def traverse(firewall, packetlevel, reportcollision=False):
    retval = advance(firewall, packetlevel, reportcollision)

    return retval, packetlevel+1


def one(dr):
    maxlevel, firewall = newfirewall(dr)
    packetlevel = 0
    severity = 0

    while packetlevel <= maxlevel:
        s, packetlevel = traverse(firewall, packetlevel)
        severity += s

    return severity


def two(dr):
    maxlevel, firewall = newfirewall(dr)
    saved = deepcopy(firewall)
    delay = 0

    while True:
        packetlevel = 0
        failed = False

        if delay:
            advance(firewall, -1)
            saved = deepcopy(firewall)

        while packetlevel <= maxlevel:
            collision, packetlevel = traverse(firewall, packetlevel, reportcollision=True)
            if collision:
                failed = True
                if delay % 10000 == 0:
                    print("failed at delay %d" % delay)
                break
        if not failed:
            return delay
        firewall = deepcopy(saved)
        delay += 1

    return delay


def main(args):
    lines = [line.strip() for line in open(args[0]).readlines()]
    dr = {}
    for line in lines:
        parts = [int(l, 10) for l in line.split(": ")]
        dr[parts[0]] = parts[1]

    print(one(dr))
    print(two(dr))


if __name__ == "__main__":
    main(sys.argv[1:])

# eof
